internal/levee.go

```diff
@@ -39,75 +41,210 @@ func run(pass *analysis.Pass) (interface{}, error) {
-					// TODO Do not hard-code logging sinks usecase
-					// TODO  Handle case of os.Stdout and os.Stderr.
-					// TODO  Do not hard-code the position of the argument, instead declaratively
-					//  specify the position of the propagated source.
-					// TODO  Consider turning propagators that take io.Writer into sinks.
-					if a := getArgumentPropagator(conf, v); a != nil {
-						sources = append(sources, source.New(a, conf))
-					} else {
-						sources = append(sources, source.New(v, conf))
-					}
-
-				case conf.IsSink(v):
-					// TODO Only variadic sink arguments are currently detected.
-					if sinkVarargs := varargs.New(v); sinkVarargs != nil {
-						for _, s := range sources {
-							if sinkVarargs.ReferredBy(s) && !s.IsSanitizedAt(v) {
-								report(pass, s, v)
-								break
-							}
-						}
-					}
+				case conf.IsSink(t):
+					graph.addSink(t, parameterNames)
+				case conf.IsPropagator(t):
+					graph.addSource(value.Name(), t)
+				case conf.IsSanitizer(t):
+					graph.addSanitizer(t)
+				default:
+					graph.addCallEdges(*t, parameterNames)
 				}
+
+			case *ssa.FieldAddr:
+				if conf.IsSourceFieldAddr(t) {
+					graph.addSource(value.Name(), t)
+				}
+
+			case *ssa.IndexAddr:
+				x, _ := t.X.(ssa.Value)
+				// example instruction: t1 = &t0[0:int]
+				// the edge needs to be inverted since we are storing
+				// a reference; in effect, t1 is now an alias for t0
+				graph.addEdge(value.Name(), x.Name())
+
+			case *ssa.MakeInterface:
+				x := t.X.(ssa.Value)
+				graph.addEdge(x.Name(), value.Name())
+
+			case *ssa.Phi:
+				for _, e := range t.Edges {
+					graph.addEdge(e.Name(), value.Name())
+				}
+
+			case *ssa.Slice:
+				x, _ := t.X.(ssa.Value)
+				graph.addEdge(x.Name(), value.Name())
+
+			case *ssa.Store:
+				graph.addEdge(t.Val.Name(), t.Addr.Name())
+
+			case *ssa.UnOp:
+				x, _ := t.X.(ssa.Value)
+				graph.addEdge(x.Name(), value.Name())
 			}
 		}
 	}
 
-	return nil, nil
+	reachabilities := graph.detectSinksReachableFromSources()
+	for _, r := range reachabilities {
+		r.report(pass)
+	}
 }
 
-func report(pass *analysis.Pass, source *source.Source, sink ssa.Node) {
-	var b strings.Builder
-	b.WriteString("a source has reached a sink")
-	fmt.Fprintf(&b, ", source: %v", pass.Fset.Position(source.Node().Pos()))
-	pass.Reportf(sink.Pos(), b.String())
+type namedNode struct {
+	name string
+	node ssa.Node
+}
+
+type sink struct {
+	name string
+	call *ssa.Call
 }
 
-func getArgumentPropagator(c *config.Config, call *ssa.Call) ssa.Node {
-	if call.Call.Signature().Params().Len() == 0 {
-		return nil
+func newSink(call *ssa.Call, i int) sink {
+	return sink{
+		name: call.Call.StaticCallee().Name() + strconv.Itoa(i),
+		call: call,
 	}
+}
 
-	firstArg := call.Call.Signature().Params().At(0)
-	if c.PropagatorArgs.ArgumentTypeRE.MatchString(firstArg.Type().String()) {
-		if a, ok := call.Call.Args[0].(*ssa.MakeInterface); ok {
-			return a.X.(ssa.Node)
+type graph struct {
```

Comment:
It's not clear to me what this new `type graph` does for us that can't be done with the existing `ssa.Node` graph.  It appears that we're reproducing a great deal of effort that's already been done for us.

More critically, this graph relies heavily on `Node.Name()` as identification.  Nodes names are not to be used semantically.  From [the `ssa` docs](https://godoc.org/golang.org/x/tools/go/ssa#Value), with focus on the third doc section:

> ```golang
> type Value interface {
>     // Name returns the name of this value, and determines how
>     // this Value appears when used as an operand of an
>     // Instruction.
>     //
>     // This is the same as the source name for Parameters,
>     // Builtins, Functions, FreeVars, Globals.
>     // For constants, it is a representation of the constant's value
>     // and type.  For all other Values this is the name of the
>     // virtual register defined by the instruction.
>     //
>     // The name of an SSA Value is not semantically significant,
>     // and may not even be unique within a function.
>     Name() string
> // ...
> ```

This graph idea is the direction we want to be heading, though.  My own thoughts were to use the existing `ssa.Node` graph, highlight those nodes which are sources, sinks, etc via independent analyzers, and then have `levee` perform a traversal of that graph to report reachability.

I'm not opposed to `putting the value graph front and center`, I just worry about reinventing a wheel.

Reply:
The concerns you raise about `Name()` are valid. I was aware of the warnings in the documentation, but given that the test cases that we have right now pass and that I didn't find anything weird or inconsistent when viewing the SSA and the value graph for a dozen or so functions in our tests, I thought it was reliable enough for us to use.

In my original investigation of Nodes, I had a hard time constructing a graph that was useful for our analysis. See e.g. [this doc](https://docs.google.com/document/d/1gp6FbYoZKqLN5AR6_1Tuqjpyd5oi5ycNrUu1FiiIEGQ/preview).

For what it's worth, the documentation for Node reads : 
```
// Node is provided to simplify SSA graph algorithms.  Clients should
// use the more specific and informative Value or Instruction
// interfaces where appropriate.
```

I'm not overly concerned about reinventing the wheel if it doesn't require much code and allows us to see our core logic more clearly. I think this is the case right now, but I may well be missing something.

Reply:
> but given that the test cases that we have right now pass and that I didn't find anything weird or inconsistent when viewing the SSA and the value graph for a dozen or so functions in our tests, I thought it was reliable enough for us to use.

Not to be glib, but our relatively simple test suite and some spot-checking is hardly reason enough to disregard a documented best practice.  But more to the point, there's no reason we need to be using strings as keys here.  Case in point...

> It seems to me like the documentation for Node suggests that in some cases using the Values might be ok?

Yes, using a `ssa.Node`, `ssa.Value`, or `ssa.Instruction` as the key to a `map` would be fine.  But at that point, the graph being produced should be the same as that presented via `Value.Referrers()` and `Instruction.Operands`.

Reply:
I believe you make valid points. Going forward I will use `Referrers/Operands`.